File: cli/cmd/ls.go
package cmd

import (
	"encoding/json"
	"fmt"

	"github.com/alperen-albayrak/kutup/cli/internal/api"
	"github.com/spf13/cobra"
)

var lsTree bool

var lsCmd = &cobra.Command{
	Use:   "ls [folder-id]",
	Short: "List files and folders",
	Args:  cobra.MaximumNArgs(1),
	RunE:  runLs,
}

func init() {
	lsCmd.Flags().BoolVar(&lsTree, "tree", false, "show full folder tree")
}

type lsEntry struct {
	ID      string  `json:"id"`
	Type    string  `json:"type"`
	Name    string  `json:"name"`
	Size    int64   `json:"size,omitempty"`
	Created string  `json:"created,omitempty"`
	Parent  *string `json:"parentId,omitempty"`
	Shared  bool    `json:"shared,omitempty"`
}

func runLs(cmd *cobra.Command, args []string) error {
	client, sess, cleanup, err := requireSessionFull()
	if err != nil {
		return err
	}
	defer cleanup()

	masterKey, err := sess.MasterKeyBytes()
	if err != nil {
		return err
	}

	cols, err := client.ListCollections()
	if err != nil {
		return err
	}
	decryptedCols := decryptCollections(cols, masterKey, sess)

	var filterParent *string
	if len(args) == 1 {
		id := args[0]
		filterParent = &id
	}

	if lsTree && filterParent == nil {
		printColTree(decryptedCols, "", 0)
		return nil
	}

	var entries []lsEntry
	for _, col := range decryptedCols {
		if filterParent == nil {
			if col.ParentCollectionID != nil {
				continue
			}
		} else {
			if col.ParentCollectionID == nil || *col.ParentCollectionID != *filterParent {
				continue
			}
		}
		entries = append(entries, lsEntry{
			ID:     col.ID,
			Type:   "folder",
			Name:   col.Name,
			Parent: col.ParentCollectionID,
			Shared: col.IsShared,
		})
	}

	if filterParent != nil {
		col := findCollection(decryptedCols, *filterParent)
		if col != nil {
			colKey, err := decryptCollectionKey(col, masterKey, sess)
			if err == nil {
				files, _ := client.ListFiles(col.ID)
				for _, f := range files {
					name, size := decryptFileMeta(&f, colKey)
					entries = append(entries, lsEntry{
						ID:      f.ID,
						Type:    "file",
						Name:    name,
						Size:    size,
						Created: formatTime(f.CreatedAt),
					})
				}
			}
		}
	}

	if jsonOut {
		enc := json.NewEncoder(cmd.OutOrStdout())
		enc.SetIndent("", "  ")
		return enc.Encode(entries)
	}

	for _, e := range entries {
		if e.Type == "folder" {
			shared := ""
			if e.Shared {
				shared = " [shared]"
			}
			fmt.Printf("📁  %-38s  %s%s\n", e.Name, e.ID, shared)
		} else {
			fmt.Printf("    %-38s  %s  %s  %s\n", e.Name, e.ID, formatBytes(e.Size), e.Created)
		}
	}
	return nil
}

func printColTree(cols []api.Collection, parentID string, depth int) {
	prefix := ""
	for i := 0; i < depth; i++ {
		prefix += "  "
	}
	for _, col := range cols {
		pID := ""
		if col.ParentCollectionID != nil {
			pID = *col.ParentCollectionID
		}
		if pID != parentID {
			continue
		}
		fmt.Printf("%s📁  %s  (%s)\n", prefix, col.Name, col.ID)
		printColTree(cols, col.ID, depth+1)
	}
}

File: crates/kutup-cli/src/commands/ls.rs
//! `kutup ls` — mirrors `cmd/ls.go`.

use anyhow::Result;
use serde::Serialize;

use crate::api::Collection;
use crate::context::require_session;
use crate::cryptohelpers::{
    decrypt_collection_key, decrypt_collections, decrypt_file_meta, find_collection,
};
use crate::output::{format_bytes, format_time};

#[derive(Serialize)]
struct LsEntry {
    id: String,
    #[serde(rename = "type")]
    entry_type: String,
    name: String,
    #[serde(skip_serializing_if = "is_zero")]
    size: i64,
    #[serde(skip_serializing_if = "String::is_empty")]
    created: String,
    #[serde(rename = "parentId", skip_serializing_if = "Option::is_none")]
    parent: Option<String>,
    #[serde(skip_serializing_if = "is_false")]
    shared: bool,
}

fn is_zero(v: &i64) -> bool {
    *v == 0
}
fn is_false(v: &bool) -> bool {
    !*v
}

pub fn run(profile: &str, json: bool, tree: bool, folder_id: Option<&str>) -> Result<()> {
    let ctx = require_session(profile)?;
    let master_key = ctx.session.master_key_bytes()?;

    let cols = ctx.client.list_collections()?;
    let cols = decrypt_collections(cols, &master_key, &ctx.session);

    let filter_parent = folder_id.map(|s| s.to_string());

    if tree && filter_parent.is_none() {
        print_col_tree(&cols, "", 0);
        return Ok(());
    }

    let mut entries: Vec<LsEntry> = Vec::new();
    for col in &cols {
        let keep = match &filter_parent {
            None => col.parent_collection_id.is_none(),
            Some(parent) => col.parent_collection_id.as_deref() == Some(parent.as_str()),
        };
        if !keep {
            continue;
        }
        entries.push(LsEntry {
            id: col.id.clone(),
            entry_type: "folder".into(),
            name: col.name.clone(),
            size: 0,
            created: String::new(),
            parent: col.parent_collection_id.clone(),
            shared: col.is_shared,
        });
    }

    if let Some(parent) = &filter_parent {
        if let Some(col) = find_collection(&cols, parent) {
            if let Ok(col_key) = decrypt_collection_key(col, &master_key, &ctx.session) {
                let files = ctx.client.list_files(&col.id).unwrap_or_default();
                for f in &files {
                    let (name, size) = decrypt_file_meta(f, &col_key);
                    entries.push(LsEntry {
                        id: f.id.clone(),
                        entry_type: "file".into(),
                        name,
                        size,
                        created: format_time(&f.created_at),
                        parent: None,
                        shared: false,
                    });
                }
            }
        }
    }

    if json {
        crate::output::print_json(&entries)?;
        return Ok(());
    }

    for e in &entries {
        if e.entry_type == "folder" {
            let shared = if e.shared { " [shared]" } else { "" };
            println!("📁  {:<38}  {}{}", e.name, e.id, shared);
        } else {
            println!(
                "    {:<38}  {}  {}  {}",
                e.name,
                e.id,
                format_bytes(e.size),
                e.created
            );
        }
    }
    Ok(())
}

fn print_col_tree(cols: &[Collection], parent_id: &str, depth: usize) {
    let prefix = "  ".repeat(depth);
    for col in cols {
        let p_id = col.parent_collection_id.as_deref().unwrap_or("");
        if p_id != parent_id {
            continue;
        }
        println!("{prefix}📁  {}  ({})", col.name, col.id);
        print_col_tree(cols, &col.id, depth + 1);
    }
}
